Feature Request: Added support for EAP dump in RADIUS Messages

aravindprasad90 opened the pull request:
Usually, for EAP related authentications, RADIUS messages contain EAP payload. Currently, RADIUS dump messages donot display the EAP frame information. Hence, added changes for displaying EAP info within RADIUS dump messages. 

Sample O/P:

1) EAP Request Identity Frame:

Before this change:
```
         EAP-Message Attribute (79), length: 19, Value: .
```
       
After this change:
```
           EAP-Message Attribute (79), length: 10, Value:
                 Response (2), id 1, len 8
                 Type Identity (1), Identity: avi
            0x0000:  0201 0008 0161 7669
```

2) EAP Request Challenge Frame:
Before this change:
```
          EAP-Message Attribute (79), length: 24, Value: ..
            0x0000:  0102 0016 0410 cd32 576a 04bd ea97 14c6
            0x0010:  9e84 ae7c 6de2
```

After this change:
```
          EAP-Message Attribute (79), length: 24, Value:
                 Request (1), id 2, len 22
                 Type MD5-challenge (4)
            0x0000:  0102 0016 0410 cd32 576a 04bd ea97 14c6
            0x0010:  9e84 ae7c 6de2
```

3) EAP Response Challenge frame:

Before this change:
```
          EAP-Message Attribute (79), length: 27, Value: ..
            0x0000:  0202 0019 0410 035f 1f16 c4e2 d71e 1553
            0x0010:  6b55 effc 0a0f 6176 69
```

After this change:
```
          EAP-Message Attribute (79), length: 27, Value:
                 Response (2), id 2, len 25
                 Type MD5-challenge (4)
            0x0000:  0202 0019 0410 035f 1f16 c4e2 d71e 1553
            0x0010:  6b55 effc 0a0f 6176 69
```

4) EAP Success Frame:

Before this change:
```
          EAP-Message Attribute (79), length: 6, Value: ..
            0x0000:  0302 0004
```

After this change:
```
          EAP-Message Attribute (79), length: 6, Value:
                 Success (3), id 2, len 4
            0x0000:  0302 0004
```

aravindprasad90 commented:
Hi Tcpdump Reviewers, 

Kindly review this pull request and let me know your views. Thanks for the support.

infrastation commented:
This work is now rebased on the current master branch and has 3 trailing whitespaces removed.

infrastation commented:
It would be nice to have a more substantial message in the proposed commit.

infrastation commented:
draft-funk-eap-ttls-v0-01 dates back to April 2007, it became RFC 5281 in August 2008. Do you know if the difference requires any changes to this code?

infrastation commented on print-radius.c:
```diff
@@ -541,6 +609,67 @@ print_attr_string(netdissect_options *ndo,
+           ND_PRINT((ndo, "\n\t\t %s (%u), id %u, len %u",
+                  tok2str(eap_code_values, "unknown", type),
+                  type, *(data+1), len));
+           length = 0;
```
Did you mean `length -= ...`?

infrastation commented on print-radius.c:
```diff
@@ -541,6 +609,67 @@ print_attr_string(netdissect_options *ndo,
+                  tok2str(eap_code_values, "unknown", type),
+                  type, *(data+1), len));
+           length = 0;
+           if (type <= 2) /* For EAP_REQUEST and EAP_RESPONSE only */
```
At this point `type` may be 0, which is not a valid value.

aravindprasad90 commented:
Hi @infrastation 

Thanks for the review. 

Addressed the following:

1) Code review comment in print-radius.c : 
>>> fixed the length variable check
>>> type 0 value check. Handled types 1,2 only. 

2) Updated the commit message

3) _"draft-funk-eap-ttls-v0-01 dates back to April 2007, it became RFC 5281 in August 2008. Do you know if the difference requires any changes to this code?"_
>>> I chcked the RFC. I think no further changes are required w r to the code additions made here. Kindly let me know if any further changes are needed here.

infrastation commented:
Thanks for the update, I cannot review right now, will update later unless another developer does it first.

aravindprasad90 commented:
@infrastation 
Hi, 
Could you kindly review this pull request

infrastation commented on print-radius.c:
```diff
@@ -541,6 +609,68 @@ print_attr_string(netdissect_options *ndo,
            length--;
         break;
+      case EAP_MESSAGE:
+           type = *data;
```
Tcpdump master branch has switched to `EXTRACT_U_1()` and `EXTRACT_S_1()` for single-byte fetching; multiple-byte fetching macros have been renamed respectively (amount of bytes instead of 1 above). Please update the code to use those new macros.

infrastation commented on print-radius.c:
```diff
@@ -541,6 +609,68 @@ print_attr_string(netdissect_options *ndo,
+      case EAP_MESSAGE:
+           type = *data;
+           len = EXTRACT_16BITS(data+2);
+           ND_PRINT((ndo, "\n\t\t %s (%u), id %u, len %u",
```
In the master branch `ND_PRINT()` is now a variadic macro, which means that instead of `ND_PRINT((ndo, x, y, z))` the code must use `ND_PRINT(x, y, z)` (no extra parentheses, no extra argument). Please update the code to reflect this change.

infrastation commented on print-radius.c:
```diff
@@ -541,6 +609,68 @@ print_attr_string(netdissect_options *ndo,
+                 case EAP_TYPE_IDENTITY:
+                      if (len - 5 > 0) {
+                         ND_PRINT((ndo, ", Identity: "));
+                         safeputs(ndo, data + 5, len - 5);
```
`safeputs()` no longer exists, please see commit e10fe19 for the replacement.

infrastation commented on print-radius.c:
```diff
@@ -541,6 +609,68 @@ print_attr_string(netdissect_options *ndo,
+
+                 case EAP_TYPE_TTLS:
+                      ND_PRINT((ndo, " TTLSv%u",
+                             EAP_TTLS_VERSION(*(data + 5)))); /* fall through */
```
There is no `break`, if this fall-through is intentional, it should be decorated with a `ND_FALL_THROUGH` and possibly a comment (see commit eebb3f4 for an example).

infrastation replied:
Have just seen the comment at the end of the line -- OK to leave as it is if you prefer.

infrastation replied:
If I am the next reviewer of this change and I fail to understand **why** the code path falls through (besides just acknowledging the fact), I may comment on that again (or add the comment myself).

infrastation commented on print-radius.c:
```diff
@@ -541,6 +609,68 @@ print_attr_string(netdissect_options *ndo,
+                      }
+                   break;
+
+                 default:
```
no-op

infrastation commented:
Thank you for waiting, I have left some quick comments, to address those you will need to rebase this change on the current master branch.

aravindprasad90 commented:
Hi @infrastation 

Thanks a lot for the review. I messed up with branch merges. 

Hence, opened a new pull request with all your review comments addressed. 
https://github.com/the-tcpdump-group/tcpdump/pull/694 

Kindly review and let me know your views.
Thanks a lot for the support.
